source/extensions/filters/network/dubbo_proxy/active_message.cc

```diff
@@ -0,0 +1,407 @@
+#include "extensions/filters/network/dubbo_proxy/active_message.h"
+
+#include "extensions/filters/network/dubbo_proxy/app_exception.h"
+#include "extensions/filters/network/dubbo_proxy/conn_manager.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace DubboProxy {
+
+// class ResponseDecoder
+ResponseDecoder::ResponseDecoder(Buffer::Instance& buffer, DubboFilterStats& stats,
+                                 Network::Connection& connection, Deserializer& deserializer,
+                                 Protocol& protocol)
+    : response_buffer_(buffer), stats_(stats), response_connection_(connection),
+      decoder_(std::make_unique<Decoder>(protocol, deserializer, this)), complete_(false) {}
+
+bool ResponseDecoder::onData(Buffer::Instance& data) {
+  ENVOY_LOG(debug, "dubbo response: the received reply data length is {}", data.length());
+
+  bool underflow = false;
+  decoder_->onData(data, underflow);
+  ASSERT(complete_ || underflow);
+  return complete_;
+}
+
+Network::FilterStatus ResponseDecoder::transportBegin() {
+  stats_.response_.inc();
+  response_buffer_.drain(response_buffer_.length());
+  ProtocolDataPassthroughConverter::initProtocolConverter(response_buffer_);
+
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus ResponseDecoder::transportEnd() {
+  if (response_connection_.state() != Network::Connection::State::Open) {
+    throw DownstreamConnectionCloseException("Downstream has closed or closing");
+  }
+
+  response_connection_.write(response_buffer_, false);
+  ENVOY_LOG(debug,
+            "dubbo response: the upstream response message has been forwarded to the downstream");
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus ResponseDecoder::messageBegin(MessageType, int64_t, SerializationType) {
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus ResponseDecoder::messageEnd(MessageMetadataSharedPtr metadata) {
+  ASSERT(metadata->message_type() == MessageType::Response ||
+         metadata->message_type() == MessageType::Exception);
+  ASSERT(metadata->response_status().has_value());
+
+  if (metadata->message_type() == MessageType::Exception) {
+    stats_.response_exception_.inc();
+  }
+
+  metadata_ = metadata;
+  switch (metadata->response_status().value()) {
+  case ResponseStatus::Ok:
+    stats_.response_success_.inc();
```

Comment:
What's the relationship between response_success_ and response_exception_ stats?

When message_type equal `MessageType::Exception`, is it possible that  response_status equal OK ?

Reply:
Response_status == OK only means that the Dubbo framework handled the request correctly. The RpcResponseType represents the result of the business layer's processing of the request. ResponseWithException means that the business layer threw an exception when processing the request, response_exception_ is used to count exceptions thrown by the business layer.

Reply:
Yes，it possible.